### ISSUE TITLE
Socks5 proxy support

### ISSUE THREAD

**Issue by cfcs** [hannesm/jackline#90]
Someone mentioned they'd want this feature for when using Tor.
I'll try to add it when I get some time.

**Comment by hannesm** [comment]
socks4a should be all we need, shouldn't it? (much simpler than user + auth overhead of socks5)

**Comment by hannesm** [comment]
also someone reported `torify` (or `torsocks` and `tor-resolv`) work fine with jackline (but I agree that a socks implementation would be great!)

**Comment by lukasepple** [comment]
Socks5 is more comonly used, isn't it`

**Comment by cfcs** [comment]
Socks4a will likely be fine for most user, however SOCKSv5 can be used by Tor to isolate circuits belonging to different contexts. I feel that is a nice enough feature that it would make sense to allow either a random username/password or a user-specified set to separate the traffic of two instances of Jackline going through the same Tor client.
[See the `IsolateSOCKSAuth` argument to the `SOCKSPort` setting.](https://www.torproject.org/docs/tor-manual.html.en#SOCKSPort)

**Comment by hannesm** [comment]
sounds good to me. I'd suggest to start with a socks library (at least client side); and afterwards finding an interface which unifies lwt sockets and socks, maybe (parts of) [conduit](https://github.com/mirage/ocaml-conduit) come in handy..

**Comment by cfcs** [comment]
Yes, let's keep this open until someone gets around to doing it :)
